### PR TITLE
Gas refunds for fake contracts

### PR DESCRIPTION
**Pull request by JcBernack**
The EVM tracks gas refunds within the implementation of the SSTORE op-code, our fake smart contracts access StateDB directly and therefore bypass this logic.

This PR replicates the gas refund mechanic of SSTORE in our methods that write to account storage.

### CODE REVIEW

**Review by akrckn — APPROVED** [review]
Seems fine to me!

File: libevm/lib/service_state_refund.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package lib
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	gethParams "github.com/ethereum/go-ethereum/params"
+)
+
+// setStateRefunds replicates the refund part of the original SSTORE gas consumption logic:
```
Comment (i-Alex) [inline libevm/lib/service_state_refund.go:9]: It seems that we implement `EIP-1283` here and it is the most actual implementation, right?
Let's add the regression test coverage using the test data from it: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1283.md#test-cases

I can see similar for the eip2200 in Geth code - `eip2200Tests`.
It is strange that they have only legacy tests there...

**Review by i-Alex — CHANGES_REQUESTED** [review]
Looks good.
Let's add some test data to check the consistency (see the comment).

File: libevm/lib/service_state_refund.go (hunk 1 above)
Comment (JcBernack) [reply libevm/lib/service_state_refund.go:9]: Not quite, 1283 was the first modification to SSTORE refunds, then came 2200, 2929 and then 3529. This implements the latest one, 3529. I will add their test cases from [EIP-3529](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-3529.md#test-cases).

Comment (i-Alex) [reply libevm/lib/service_state_refund.go:9]: Ok, thanks. than 3529

Comment (JcBernack) [reply libevm/lib/service_state_refund.go:9]: done

**Review by i-Alex — APPROVED** [review]

**Pull request merged**